src/odin/maneuversbuilder.cc

```diff
@@ -2810,55 +2813,97 @@ uint16_t ManeuversBuilder::GetExpectedTurnLaneDirection(Maneuver& maneuver) cons
 }
 
 void ManeuversBuilder::ProcessTurnLanes(std::list<Maneuver>& maneuvers) {
+  auto prev_man = maneuvers.begin();
   auto curr_man = maneuvers.begin();
   auto next_man = maneuvers.begin();
 
+  // Set current maneuver
+  if (next_man != maneuvers.end()) {
+    ++next_man;
+    curr_man = next_man;
+  }
+
+  // Set next maneuver
   if (next_man != maneuvers.end()) {
     ++next_man;
   }
 
   // Walk the maneuvers to activate turn lanes
-  while (next_man != maneuvers.end()) {
+  while (curr_man != maneuvers.end()) {
 
     // Only process driving maneuvers
     if (curr_man->travel_mode() == TripLeg_TravelMode::TripLeg_TravelMode_kDrive) {
 
-      // Keep track of the remaining step distance in kilometers
-      float remaining_step_distance = curr_man->length();
-
       // Walk maneuvers by node (prev_edge of node has the turn lane info)
       // Assign turn lane at transition point
       auto prev_edge = trip_path_->GetPrevEdge(curr_man->begin_node_index());
       if (prev_edge && (prev_edge->turn_lanes_size() > 0)) {
         // If not a short fork then process for turn lanes
-        if (!((remaining_step_distance < kShortForkThreshold) &&
+        if (!((curr_man->length() < kShortForkThreshold) &&
               ((curr_man->type() == DirectionsLeg_Maneuver_Type_kStayLeft) ||
                (curr_man->type() == DirectionsLeg_Maneuver_Type_kStayRight) ||
                (curr_man->type() == DirectionsLeg_Maneuver_Type_kStayStraight)))) {
-          prev_edge->ActivateTurnLanes(GetExpectedTurnLaneDirection(*(curr_man)),
-                                       remaining_step_distance, curr_man->type(), next_man->type());
+          prev_edge->ActivateTurnLanes(GetExpectedTurnLaneDirection(prev_edge, *(curr_man)),
+                                       curr_man->length(), curr_man->type(), next_man->type());
         }
       }
 
-      // Assign turn lanes within step
-      for (auto index = (curr_man->begin_node_index() + 1); index < curr_man->end_node_index();
-           ++index) {
+      // Track whether there is an intermediate traversable intersecting edge in the same
+      // direction as the next maneuver
+      bool has_directional_intersecting_edge = false;
+
+      // Keep track of the remaining step distance in kilometers
+      float remaining_step_distance = 0.f;
+      // Add the prev_edge length as we skip it in the loop below
+      if (prev_edge) {
+        remaining_step_distance += prev_edge->length();
+      }
+
+      // Assign turn lanes within step, walking backwards from end to begin node
+      for (auto index = (prev_man->end_node_index() - 1); index > prev_man->begin_node_index();
```

Comment:
nit: instead of `auto` specify the type, in this case, `uint32_t`. I don't know if its possible due to protections up the call-chain, but if `prev_man->end_node_index()` is ever 0, bad juju could happen. Having the type explicit here helps in highlighting the fact that we _could_ have a very large value if the end index was somehow 0.

(2️⃣¢)

Reply:
Fixed :+1: